File: Week 6/3. Motion type/ros2_ws/src/motion_types/src/linear.cpp
/*
Turtle to move linearly based on specified direction (forward or backward), speed, and distance.

Author: Abisoye Akinloye
Date: September 2022
*/

#include "rclcpp/rclcpp.hpp"
#include "geometry_msgs/msg/twist.hpp"

class MoveLinearly : public rclcpp::Node
{
public:
    MoveLinearly() : Node("linear_movement")
    {
        MoveLinearly::set_movement_params();
        velocity_publisher_ = this->create_publisher<geometry_msgs::msg::Twist>("turtle1/cmd_vel", 10);
        timer_ = this->create_wall_timer(std::chrono::seconds(1), std::bind(&MoveLinearly::move_linear, this));
    }

private:
    void set_movement_params()
    {
        std::cout << "Enter the desired distance (m): ";
        std::cin >> distance;
        std::cout << "Enter the desired velocity (m/s): ";
        std::cin >> velocity;
        std::cout << "Enter [1] for forward and [0] for backward direction: ";
        std::cin >> isForward;

        t0 = time.now().seconds();
    }

    void move_linear()
    {
        t1 = time.now().seconds();
        int dt = t1 - t0;
        RCLCPP_INFO(this->get_logger(), "Change in time: %i", dt);

        displacement = velocity * dt;
        if (displacement < distance)
        {
            if (isForward)
                vel_msg.linear.x = abs(velocity);
            else
                vel_msg.linear.x = -abs(velocity);
        }
        else
        {
            vel_msg.angular.z = 0.0;
            RCLCPP_INFO(this->get_logger(), "Specified distance of %.1f reached.", distance);
            timer_->cancel();
        }

        velocity_publisher_->publish(vel_msg);
    }

    rclcpp::Publisher<geometry_msgs::msg::Twist>::SharedPtr velocity_publisher_;
    rclcpp::TimerBase::SharedPtr timer_;
    geometry_msgs::msg::Twist vel_msg;
    rclcpp::Clock time;
    float distance, displacement, velocity;
    bool isForward; int t1, t0;
};

int main(int argc, char *argv[])
{
    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<MoveLinearly>());
    rclcpp::shutdown();

    return 0;
}

File: Week 6/3. Motion type/ros2_ws/install/motion_types/lib/motion_types/linear.py
#!/usr/bin/env python3

'''
Turtle to move linearly based on specified direction (forward or backward), speed, and distance.

Author: Abisoye Akinloye
Date: September 2022
'''
import rclpy
from rclpy.node import Node
from geometry_msgs.msg import Twist


class MoveLinearly(Node):
    def __init__(self):
        super().__init__("linear_movement")
        self.velocity_publisher_ = self.create_publisher(
            Twist, "/turtle1/cmd_vel", 10)
        self.timer_ = self.create_timer(0.05, self.move_callback)
        # self.get_logger().info("The turtle is about to move linearly")
        self.distance, self.velocity, self.direction = self.set_params()

    def set_params(self):
        '''
        Func prompt to input distance, speed, and direction.
        '''
        distance = float(input("Enter the desired distance (m): "))
        velocity = float(input("Enter the desired velocity (m/s): "))
        direction = int(
            input("Enter [1] for forward and [0] for backward direction: "))
        self.t0 = self.get_clock().now().seconds_nanoseconds()[
            0]  # initial time
        return distance, velocity, direction

    def move_callback(self):
        cmd_vel = Twist()
        t1 = self.get_clock().now().seconds_nanoseconds()[0]
        dt = t1-self.t0  # change in time

        # displacement because it is a distance taken in a specified direction
        displacement = self.velocity * dt

        if displacement < self.distance:
            if self.direction == 1:
                cmd_vel.linear.x = abs(self.velocity)
            elif self.direction == 0:
                cmd_vel.linear.x = -abs(self.velocity)
            else:
                print("Invalid input!")
                rclpy.shutdown()
        else:
            cmd_vel.linear.x = 0.0
            self.get_logger().info("Distance reached!!!")
            self.timer_.cancel()

        self.velocity_publisher_.publish(cmd_vel)


def main(args=None):
    rclpy.init(args=args)
    move = MoveLinearly()
    rclpy.spin(move)
    rclpy.shutdown()


if __name__ == '__main__':
    main()
